### PR TITLE
Hashcat encryption

### PR DESCRIPTION
**Pull request by kkowalski-reef**
- Encrypts the first stage of the hashcat synthetic job
- Uses miner hotkey as part of the key for the first stage

### CODE REVIEW

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            return file.read()
+
+    @property
+    def answer(self) -> str:
+        return self._hash(
+            "".join(["".join(passwords) for passwords in self.passwords]).encode("utf-8")
+        ).decode("utf-8")
+
+    def __str__(self) -> str:
+        return f"V2SyntheticJob {self.algorithms} {self.params}"
```
Comment (agoncharov-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py:133]:
```suggestion
        return f"{self.__class__.__name__} {self.algorithms} {self.params}"
```

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            "num_digits": [params.num_digits for params in self.params],
+        }
+        return pickle.dumps(data)
+
+    def docker_run_cmd(self) -> list[str]:
+        return ["/script.py", self.first_password]
+
+    def raw_script(self) -> str:
+        with open(Path(__file__).parent / "v2_decrypt.py") as file:
+            return file.read()
```
Comment (agoncharov-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py:124]:
```suggestion
        return (Path(__file__).parent / "v2_decrypt.py").read_text()
```

**Review by agoncharov-reef — APPROVED** [review]
I think copy-paste for new versions of models is totally OK. No dependencies is good.

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            _passwords = set()
+            while len(_passwords) < _params.num_hashes:
+                _passwords.add(
+                    cls.random_string(
+                        num_letters=_params.num_letters, num_digits=_params.num_digits
+                    )
+                )
+            passwords.append(sorted(list(_passwords)))
+
+        first_password = f"{miner_hotkey}-{cls.random_string(num_letters=48, num_digits=0)}"
```
[the remark this answers is not in the document]
Comment (mzukowski-reef) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/v2_synthetic_job.py:58]: num_digits is only for hashcat - so we have better control over password space - meaning - breaking password with additional letter is much harder than with additional digit or two - this password is different - it is just a random string - 48 letters makes it unbreakable :)

**Pull request merged**